CIP-????/README.md

```diff
@@ -0,0 +1,471 @@
+(case t cs)
+
+-- Constructs a product with the given fields
+(prod ts)
+-- Accesses the i'th field of the given product
+(index i t)
+```
+
+This is cleaner in some ways, and was the first prototype we implemented, but it has the following problems:
+- It adds two more term constructors to the language
+- It is significantly slower in practice, because the combined operation of processing a sum-of-products is extremely common
+
+#### 2. Bind Case Variables in the Syntax for Case Branches
+
+The design presented here allows the case branches to be arbitrary terms, which must be evaluated and then applied to the fields.
+
+A more complex design would be to have the bindings for the variables be part of the case expression instead. That is, the current design does this:
+
+```
+(case (constr 1 a b) (\x -> t1) (\x y -> t2))
+```
+(the literal lambdas here could be arbitrary terms)
+
+whereas we could do this:
+
+```
+(case (constr 1 x y) (alt x t1) (alt x y t2))
+```
+
+The effect of this is that we know statically that we have a function, and we can jump right into evaluating the _body_ of the case branch instead of first having to evaluate it to a function and apply the arguments.
+
+This saves us a small but meaningful amount of overhead at the cost of making the implementation significantly more complex.
+
+We prototyped this version and it was noticeably faster (~10%).
+However, performance investigations showed that we can realize a significant amount of the performance gain through other means, leaving only about 3% unclaimed.[^see-appendix-2]
+We think that this is an acceptable cost for a simpler implementation.
+
+[^see-appendix-2]: See Appendix 2 for more details.
+
+## Path to Active
+
+### Acceptance Criteria
+
+- [ ] Implementation in `plutus`, including specification and costing of the new operations
+- [ ] Further benchmarking with additional real-world examples
+- [ ] `plutus` changes integrated into `cardano-ledger`, including specification
+
+### Implementation Plan
+
+This plan will be implemented by Michael Peyton Jones and the Plutus Core team with assistance from the Ledger team.
+
+A prototype implementation already exists and is functional.
+
+## Appendices
+
+### Appendix 1: Changes to Typed Plutus Core
+
+While Typed Plutus Core is not part of the specification of Cardano, it is still interesting and informative to give the changes here.
+Plutus Core was originally conceived of as a typed language, and making sure that we can express the changes cleanly in a typed setting means that we ensure that the semantics make sense and that it will continue to be easy to compile to Plutus Core from typed languages.
+
+We add the following new type constructors to the type language of Typed Plutus Core:
+```
+(sum tys)
+(prod tys)
+```
+These correspond to sum and product types.
+While we are not using sums-and-products at the term level, we can keep it at the type level since that does not affect evaluation.
```

Comment:
I guess there's no real reason not to do it. They always appear nested anyway. It just seemed slightly simpler to me two nested n-ary types rather than one which has to have a list of lists of children...

Reply:
> It just seemed slightly simpler to me two nested n-ary types rather than one which has to have a list of lists of children...

Sure, but if you give me that, I'm going to use it and then every time someone asks why `prod`, which is supposed to be `unit`, used as `void`, I will tell them to ask you that! And once I get away with that, I'll start thinking "we have this nice inert type primitive, can I find a use for it?" and then I'll realize that I can have a convention that `prod a` "computes" to `a` when it comes to runtime type tags (but not from the type checker's point of view), allowing me to encode `Proxy a` as `[prod a]` (both have only one inhabitant, since `prod a` is uninhabitable thus only leaving the empty list as a single possible value of `[prod a]`, which is the value I can attach a type for `a` to) within the type signature of a builtin and then I'll bring it up to the team just to be an ass and after a hundred emails discussion I'll get fired for wasting everyone's time -- you see how complicated things get when one chooses convenience over coherent semantics?

Reply:
Alright, I'll change it :)

Reply:
You know what... I'm taking that back. I would use `prod` as `void` if we didn't have `void`, but this proposal also allows empty sum types and that is a much better option. It would be very tempting for me to utilize `prod` in builtins the way I described above, but I promise to resist the temptation. I still think it's a good idea not to add to a language something that you don't need, because who knows how it may backfire, but I don't have anything concrete against the specific choice, only vague feelings. Sorry for the noise!